runtime/src/validate_block/implementation.rs

```diff
@@ -0,0 +1,230 @@
+	let storage = Storage::<B>::new(
+		block_data.witness_data,
+		block_data.witness_data_storage_root,
+	).expect("Witness data and storage root always match; qed");
+
+	let _guard = unsafe {
+		STORAGE = Some(Box::new(storage));
+		(
+			// Replace storage calls with our own implementations
+			rio::ext_get_allocated_storage.replace_implementation(ext_get_allocated_storage),
+			rio::ext_get_storage_into.replace_implementation(ext_get_storage_into),
+			rio::ext_set_storage.replace_implementation(ext_set_storage),
+			rio::ext_exists_storage.replace_implementation(ext_exists_storage),
+			rio::ext_clear_storage.replace_implementation(ext_clear_storage),
+			rio::ext_storage_root.replace_implementation(ext_storage_root),
+		)
+	};
+
+	E::execute_block(block);
+}
+
+/// The storage implementation used when validating a block.
+struct Storage<B: BlockT> {
+	witness_data: MemoryDB<<HashingOf<B> as HashT>::Hasher>,
+	overlay: hashbrown::HashMap<Vec<u8>, Option<Vec<u8>>>,
+	storage_root: HashOf<B>,
+}
+
+impl<B: BlockT> Storage<B> {
+	/// Initialize from the given witness data and storage root.
+	///
+	/// Returns an error if given storage root was not found in the witness data.
+	fn new(
+		data: WitnessData,
+		storage_root: HashOf<B>
+	) -> Result<Self, &'static str> {
+		let mut db = MemoryDB::default();
+		data.into_iter().for_each(|i| { db.insert(&[], &i); });
+
+		if !db.contains(&storage_root, &[]) {
+			return Err("Witness data does not contain given storage root.")
+		}
+
+		Ok(Self {
+			witness_data: db,
+			overlay: Default::default(),
+			storage_root,
+		})
+	}
+}
+
+impl<B: BlockT> StorageT for Storage<B> {
+	fn get(&self, key: &[u8]) -> Option<Vec<u8>> {
+		self.overlay.get(key).cloned().or_else(|| {
+			read_trie_value(
+				&self.witness_data,
+				&self.storage_root, key
+			).ok()
+		}).unwrap_or(None)
+	}
+
+	fn insert(&mut self, key: &[u8], value: &[u8]) {
+		self.overlay.insert(key.to_vec(), Some(value.to_vec()));
+	}
+
+	fn remove(&mut self, key: &[u8]) {
+		self.overlay.insert(key.to_vec(), None);
+	}
+
+	fn storage_root(&mut self) -> [u8; STORAGE_ROOT_LEN] {
+		let root = match delta_trie_root(
+			&mut self.witness_data,
+			self.storage_root.clone(),
+			self.overlay.drain()
+		) {
+			Ok(root) => root,
+			Err(_) => return [0; STORAGE_ROOT_LEN],
+		};
+
+		assert!(root.as_ref().len() <= STORAGE_ROOT_LEN);
+		let mut res = [0; STORAGE_ROOT_LEN];
+		res.copy_from_slice(root.as_ref());
+		res
+	}
+}
+
+unsafe fn ext_get_allocated_storage(
+	key_data: *const u8,
+	key_len: u32,
+	written_out: *mut u32,
+) -> *mut u8 {
+	let key = slice::from_raw_parts(key_data, key_len as usize);
+	match STORAGE.as_mut().expect(STORAGE_SET_EXPECT).get(key) {
+		Some(mut value) => {
+			*written_out = value.len() as u32;
+			let ptr = value.as_mut_ptr();
+			mem::forget(ptr);
+			ptr
+		},
+		None => {
+			*written_out = u32::max_value();
+			ptr::null_mut()
+		}
+	}
+}
+
+unsafe fn ext_set_storage(
```

Comment:
i would prefer to extract these to a submodule where we can be a bit more sure about invariants

Reply:
Hmm, this is already its own module and you are sure about the invariants. This function is only usable by `validate_block`. I would not put this into another file.